From 58f366ae608faa7f117178d3f918f62598f0815f Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Tue, 22 Sep 2020 12:24:11 -0700
Subject: [PATCH 1636/1852] v4l: xilinx: vpss-scaler: Fix warnings for
 NULL_RETURNS

commit afa4ed26ab0f89de20b452d3b8806f31616baf68 from
https://github.com/Xilinx/linux-xlnx.git

This patch fix the warnings for the NULL_RETURNS type.

The actual warnings are:

For __xscaler_get_pad_format():
Event return_null: 	Explicitly returning null

For xscaler_get_format():
Event returned_null: 	"__xscaler_get_pad_format" returns "NULL".
Event dereference: 	Dereferencing "__xscaler_get_pad_format(xscaler,
cfg, fmt->pad, fmt->which)", which is known to be "NULL".

For xscaler_set_format():
Event returned_null: 	"__xscaler_get_pad_format" returns
"NULL".
Event var_assigned: 	Assigning: "format" = "NULL" return value from
"__xscaler_get_pad_format".
Event dereference: 	Dereferencing "format", which is known to be
"NULL".

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../platform/xilinx/xilinx-vpss-scaler.c      | 27 ++++++++++++++-----
 1 file changed, 21 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-vpss-scaler.c b/drivers/media/platform/xilinx/xilinx-vpss-scaler.c
index e278c04edf9d..939c8b4b0ce8 100644
--- a/drivers/media/platform/xilinx/xilinx-vpss-scaler.c
+++ b/drivers/media/platform/xilinx/xilinx-vpss-scaler.c
@@ -1701,15 +1701,22 @@ __xscaler_get_pad_format(struct xscaler_device *xscaler,
 			 struct v4l2_subdev_pad_config *cfg,
 			 unsigned int pad, u32 which)
 {
+	struct v4l2_mbus_framefmt *format;
+
 	switch (which) {
 	case V4L2_SUBDEV_FORMAT_TRY:
-		return v4l2_subdev_get_try_format(&xscaler->xvip.subdev, cfg,
-						  pad);
+		format = v4l2_subdev_get_try_format(&xscaler->xvip.subdev, cfg,
+						    pad);
+		break;
 	case V4L2_SUBDEV_FORMAT_ACTIVE:
-		return &xscaler->formats[pad];
+		format = &xscaler->formats[pad];
+		break;
 	default:
-		return NULL;
+		format = NULL;
+		break;
 	}
+
+	return format;
 }
 
 static int xscaler_get_format(struct v4l2_subdev *subdev,
@@ -1717,9 +1724,14 @@ static int xscaler_get_format(struct v4l2_subdev *subdev,
 			      struct v4l2_subdev_format *fmt)
 {
 	struct xscaler_device *xscaler = to_scaler(subdev);
+	struct v4l2_mbus_framefmt *format;
+
+	format = __xscaler_get_pad_format(xscaler, cfg, fmt->pad,
+					  fmt->which);
+	if (!format)
+		return -EINVAL;
 
-	fmt->format = *__xscaler_get_pad_format(xscaler, cfg, fmt->pad,
-						fmt->which);
+	fmt->format = *format;
 
 	return 0;
 }
@@ -1732,6 +1744,9 @@ static int xscaler_set_format(struct v4l2_subdev *subdev,
 	struct v4l2_mbus_framefmt *format;
 
 	format = __xscaler_get_pad_format(xscaler, cfg, fmt->pad, fmt->which);
+	if (!format)
+		return -EINVAL;
+
 	*format = fmt->format;
 
 	format->width = clamp_t(unsigned int, fmt->format.width,
-- 
2.31.1

